ionc/ion_timestamp.c

```diff
@@ -930,26 +1031,42 @@ iERR ion_timestamp_equals(const ION_TIMESTAMP *ptime1, const ION_TIMESTAMP *ptim
         goto is_false;
     }
 
-    IONCHECK(ion_timestamp_get_precision(ptime1, &precision1));
-    IONCHECK(ion_timestamp_get_precision(ptime2, &precision2));
-    if (precision1 != precision2) {
-        goto is_false;
-    }
-
-    IONCHECK(ion_timestamp_to_time_t(ptime1, &t1));
-    IONCHECK(ion_timestamp_to_time_t(ptime2, &t2));
-    if (t1 != t2) {
-        goto is_false;
+    if (instant_only) {
+        // Fractional precision does not matter in this case. Trim any trailing zeros from the fractions, then compare.
+        decQuadReduce(&fraction_trimmed1, &ptime1->fraction, pcontext);
+        decQuadReduce(&fraction_trimmed2, &ptime2->fraction, pcontext);
+        ion_decimal_equals(&fraction_trimmed1, &fraction_trimmed2, pcontext, &decResult);
+        if (!decResult) goto is_false;
+        IONCHECK(_ion_timestamp_to_utc(ptime1, &ptime1_compare));
+        IONCHECK(_ion_timestamp_to_utc(ptime2, &ptime2_compare));
     }
-
-    if (precision1 == ION_TS_FRAC) {
-        decQuadCompare(&decResult, &(ptime1->fraction), &(ptime2->fraction), pcontext);
-        if (decQuadIsZero(&decResult)) {
-            goto is_true;
+    else {
+        IONCHECK(ion_timestamp_get_precision(ptime1, &precision1));
+        IONCHECK(ion_timestamp_get_precision(ptime2, &precision2));
+        if (precision1 != precision2) {
+            goto is_false;
         }
-        else {
+
+        if (HAS_TZ_OFFSET(ptime1) ^ HAS_TZ_OFFSET(ptime2)) {
             goto is_false;
         }
+        if (precision1 == ION_TS_FRAC) { // Then precision2 is also ION_TS_FRAC.
+            ion_decimal_equals(&(ptime1->fraction), &(ptime2->fraction), pcontext, &decResult);
+            if (!decResult) goto is_false;
+        }
+        ptime1_compare = *ptime1;
+        ptime2_compare = *ptime2;
+    }
+
+    if    ((ptime1_compare.year != ptime2_compare.year)
```

Comment:
I suspect all these nested parentheses aren't needed;  not a big deal either way